teps/0090-looping.md

```diff
@@ -0,0 +1,429 @@
+      I'd prefer the concept of a single Task does a single thing, rather than overloading it like this. Especially 
+      when viewing it in the dashboard etc, things get lost" ~ [bitsofinfo][kaniko-example-2]
+
+We need to address these challenges and limitations to improve the composability, scalability, flexibility,
+reusability and debuggability of *Tekton Pipelines*. 
+
+**In this TEP, we aim to provide a way to run the same `Task` or `Custom Task` with varying `Parameters` by spinning up 
+a `TaskRun` or `Run` for each `Parameter` in a loop. This looping construct is aimed at improving the composability, 
+scalability, flexibility and reusability of *Tekton Pipelines***.
+
+### Goals
+
+- Executing `Tasks` and `Custom Tasks` in a loop with varying `Parameter` values.
+- Configuring whether the `TaskRuns` and `Runs` created in the loop execute sequentially or parallelly.
+
+### Non-Goals
+
+- Terminating early when the `Tasks` or `Custom Tasks` are executed parallely - in-progress `TaskRuns` and `Runs` have 
+to complete execution before termination.
+- Ignoring a failure when the `Tasks` or `Custom Tasks` are executed sequentially - addressed in [TEP-0050][tep-0050].
+- Controlling the concurrency of `TaskRuns` or `Runs` created in a given loop - to be addressed more broadly in Tekton
+Pipelines ([tektoncd/pipeline: issue#2591][issue-2591], [tektoncd/experimental: issue#804][issue-804]).
+
+### Use Cases
+
+#### Parallel Kaniko Build
+
+As a `Pipeline` author, I [need to build multiple images][kaniko-example-1] from one repository using the same `Task`. 
+I choose to use the [*kaniko*][kaniko-task] `Task` from the *Tekton Catalog*. Let's assume it's three images. I want to 
+pass in varying `Parameter` values for `IMAGE` to create three `TaskRuns`, one for each image. 
+
+```
+                                     clone
+                                       |
+                                       v
+                 --------------------------------------------------
+                   |                   |                       |
+                   v                   v                       v   
+            ko-build-image-1     ko-build-image-2        ko-build-image-3     
+```
+
+In other circumstances, the `Parameter` values for `IMAGE` may be produced by a previous `Task` in the `Pipeline` 
+instead of supplying them myself.
+
+Read more in [user experience report #1][kaniko-example-1] and [user experience report #2][kaniko-example-2].
+
+#### Dynamic Parallel Docker Build 
+
+As a `Pipeline` author, I have several dockerfiles in my repository. 
+
+```
+/ docker / Dockerfile
+  python / Dockerfile
+  Ubuntu / Dockerfile
+...
+```
+
+I have a *clone* `Task` that fetches the repository to a shared `Workspace`. Then I have a *get-dir* `Task` that 
+produces a `Result` array with the directory names of the dockerfiles. Finally, I want to dynamically generate the 
```

Comment:
> I think it's a tricky distinction. I'd like to prevent the case where a TaskRun can generate infinite subsequent TaskRuns, but even there there's a distinction: if a TaskRun generates N outputs that each feed into a TaskRun, and that N is bounded (either explicitly by some limit we impose, or just implicitly because you have to count to N before proceeding), I think it could be okay.
> 
> If a TaskRun is dynamically emitting results that are producing new TaskRuns _as results are emitted_, you can get into a situation where it's impossible to tell whether the feed will ever end, because you don't know whether the emitting end will ever stop. That's the more important one to avoid.
> 
> (I _think_ we're discussing the former scenario, which I think could be fine. Please let me know if I'm misunderstanding.)

@imjasonh yes, we are discussing the former scenario because the a task cannot be executed before its parent task is done -- therefore it's not possible to have a feed of results that generate results that may never stop

to clarify, `dynamic` here refers to the fact that by using results in the matrix, different executions of a given pipeline may generate different number of taskruns from that matrix 

also, added a requirement that users can configure the maximum number of taskruns generated from matrix, with a default value provided